accounts/abi/argument.go

```diff
@@ -152,27 +148,54 @@ func (arguments Arguments) unpackTuple(v interface{}, output []byte) error {
 }
 
 // unpackAtomic unpacks ( hexdata -> go ) a single value
-func (arguments Arguments) unpackAtomic(v interface{}, output []byte) error {
-	// make sure the passed value is arguments pointer
-	valueOf := reflect.ValueOf(v)
-	if reflect.Ptr != valueOf.Kind() {
-		return fmt.Errorf("abi: Unpack(non-pointer %T)", v)
-	}
-	arg := arguments[0]
-	if arg.Indexed {
-		return fmt.Errorf("abi: attempting to unpack indexed variable into element.")
+func (arguments Arguments) unpackAtomic(v interface{}, marshalledValues []interface{}) error {
+
+	if len(marshalledValues) != 1 {
+		return fmt.Errorf("abi: wrong length, expected single value, got %d", len(marshalledValues))
 	}
 
-	value := valueOf.Elem()
+	elem := reflect.ValueOf(v).Elem()
+	reflectValue := reflect.ValueOf(marshalledValues[0])
+	return set(elem, reflectValue, arguments.NonIndexed()[0])
+}
 
-	marshalledValue, err := toGoType(0, arg.Type, output)
-	if err != nil {
-		return err
+// UnpackValues can be used to unpack ABI-encoded hexdata according to the ABI-specification,
+// without supplying a struct to unpack into. Instead, this method returns a list containing the
+// values. An atomic argument will be a list with one element.
+func (arguments Arguments) UnpackValues(data []byte) ([]interface{}, error) {
+
+	retval := make([]interface{}, 0, arguments.LengthNonIndexed())
+
+	virtualArgs := 0
+
+	for index, arg := range arguments.NonIndexed() {
+
+		marshalledValue, err := toGoType((index+virtualArgs)*32, arg.Type, data)
+
+		if arg.Type.T == ArrayTy {
+			//If we have a static array, like [3]uint256, these are coded as
+			// just like uint256,uint256,uint256.
+			// This means that we need to add two 'virtual' arguments when
+			// we count the index from now on
+
+			virtualArgs += arg.Type.Size - 1
+		}
+
+		if err != nil {
+			return nil, err
+		}
+		retval = append(retval, marshalledValue)
 	}
-	return set(value, reflect.ValueOf(marshalledValue), arg)
+	return retval, nil
+}
+
+// UnpackValues performs the operation Go format -> Hexdata
+// It is the semantic opposite of UnpackValues
+func (arguments Arguments) PackValues(args []interface{}) ([]byte, error) {
```

Comment:
Why do we need this method, if we can just call `Pack(args...)` instead?

Reply:
We don't _need_ it, I just added to have coder<->decoder pipe with identical input<->output format. LMK if I should remove it.